tracer/src/Datadog.Trace.ClrProfiler.Native/rejit_handler.cpp

```diff
@@ -618,6 +618,125 @@ void RejitHandler::RequestRejitForNGenInliners()
+                  "(", (integration.target_method.signature_types.size() - 1), " params)' method implementation.");
+
+    // Now we enumerate all methods with the same target method name. (All overloads of the method)
+    auto enumMethods = Enumerator<mdMethodDef>(
+        [&metadataImport, &integration, typeDef](HCORENUM* ptr, mdMethodDef arr[], ULONG max, ULONG* cnt) -> HRESULT {
+            return metadataImport->EnumMethodsWithName(ptr, typeDef, integration.target_method.method_name.c_str(), arr,
+                                                       max, cnt);
+        },
+        [&metadataImport](HCORENUM ptr) -> void { metadataImport->CloseEnum(ptr); });
+
+    auto enumIterator = enumMethods.begin();
+    for (; enumIterator != enumMethods.end(); enumIterator = ++enumIterator)
+    {
+        auto methodDef = *enumIterator;
+
+        // Extract the function info from the mdMethodDef
+        const auto caller = GetFunctionInfo(metadataImport, methodDef);
+        if (!caller.IsValid())
+        {
+            Logger::Warn("    * The caller for the methoddef: ", TokenStr(&methodDef), " is not valid!");
+            continue;
+        }
+
+        // We create a new function info into the heap from the caller functionInfo in the stack, to
+        // be used later in the ReJIT process
+        auto functionInfo = FunctionInfo(caller);
+        auto hr = functionInfo.method_signature.TryParse();
+        if (FAILED(hr))
+        {
+            Logger::Warn("    * The method signature: ", functionInfo.method_signature.str(), " cannot be parsed.");
+            continue;
+        }
+
+        // Compare if the current mdMethodDef contains the same number of arguments as the
+        // instrumentation target
+        const auto numOfArgs = functionInfo.method_signature.NumberOfArguments();
+        if (numOfArgs != integration.target_method.signature_types.size() - 1)
+        {
+            Logger::Debug("    * The caller for the methoddef: ", caller.name,
+                          " doesn't have the right number of arguments (", numOfArgs, " arguments).");
+            continue;
+        }
+
+        // Compare each mdMethodDef argument type to the instrumentation target
+        bool argumentsMismatch = false;
+        const auto methodArguments = functionInfo.method_signature.GetMethodArguments();
+
+        Logger::Debug("    * Comparing signature for method: ", caller.type.name, ".",
+                      caller.name);
+        for (unsigned int i = 0; i < numOfArgs; i++)
+        {
+            const auto argumentTypeName = methodArguments[i].GetTypeTokName(metadataImport);
+            const auto integrationArgumentTypeName = integration.target_method.signature_types[i + 1];
+            Logger::Debug("        -> ", argumentTypeName, " = ", integrationArgumentTypeName);
+            if (argumentTypeName != integrationArgumentTypeName && integrationArgumentTypeName != WStr("_"))
```

Comment:
I may be misunderstanding but writing anyway.
`GetTypeTokName` does not return a different name depending on whether the parameter is passed by ref or not. So there can be two methods with the same number of parameters and with the same names and the difference is only in by ref. Is it can cause error here?

Reply:
Well In fact `GetTypeTokName` calls `GetSigTypeTokName` and handles that, the name has a `&` suffix in case is a ref.

Here is the code: https://github.com/DataDog/dd-trace-dotnet/blob/e591eb9e39d032dfa9164b069af3e0c854bd19ec/tracer/src/Datadog.Trace.ClrProfiler.Native/clr_helpers.cpp#L603-L606

Reply:
By running an example you can see in the native logs the `&` char when is by ref:

```
12/03/21 06:16:18.922 PM [572332|536824] [debug]   Looking for 'CallTargetNativeTest.WithRefArguments.VoidRefMethod(2 params)' method implementation.
12/03/21 06:16:18.922 PM [572332|536824] [debug]     * Comparing signature for method: CallTargetNativeTest.WithRefArguments.VoidRefMethod
12/03/21 06:16:18.922 PM [572332|536824] [debug]         -> System.String& = _
12/03/21 06:16:18.922 PM [572332|536824] [debug]         -> System.Int32& = _
12/03/21 06:16:18.922 PM [572332|536824] [debug]     * Enqueue for ReJIT [ModuleId=140713391636808, MethodDef=0600011c, AppDomainId=1845728086960, Assembly=CallTargetNativeTest, Type=CallTargetNativeTest.WithRefArguments, Method=VoidRefMethod(2 params), Signature=200201100e1008]
```

and without it:
```
12/03/21 06:16:18.920 PM [572332|536824] [debug]   Looking for 'CallTargetNativeTest.With2Arguments.VoidMethod(2 params)' method implementation.
12/03/21 06:16:18.920 PM [572332|536824] [debug]     * Comparing signature for method: CallTargetNativeTest.With2Arguments.VoidMethod
12/03/21 06:16:18.920 PM [572332|536824] [debug]         -> System.String = _
12/03/21 06:16:18.920 PM [572332|536824] [debug]         -> System.Int32 = _
12/03/21 06:16:18.920 PM [572332|536824] [debug]     * Enqueue for ReJIT [ModuleId=140713391636808, MethodDef=06000064, AppDomainId=1845728086960, Assembly=CallTargetNativeTest, Type=CallTargetNativeTest.With2Arguments, Method=VoidMethod(2 params), Signature=2002010e08]
```

Reply:
Thanks, great!